### ISSUE TITLE
gathering asyncio open_connection ends up with timing gaps

### ISSUE THREAD

**Issue by CastixGitHub** [python/cpython#100419]
description at bottom (ok it may seem a public webserver scanner, but that's the simpler reproducible example to reason about, our purpose is different)
```python
# © 2022 Vincenzo Castiglia <vincenzo.castiglia@par-tec.it> 3F2AD937F3D77362
import asyncio
import socket
import errno
import random
import struct
import sys
from time import time


STATUSES = {
    0: 'OK',
    errno.ETIMEDOUT: 'TIMEDOUT',
    errno.ECONNREFUSED: 'REFUSED',
    errno.EHOSTUNREACH: 'NOROUTE',
    errno.ENETUNREACH: 'UNREACHABLE',
    errno.ECONNRESET: 'CONNRESET',
    # errno.ECONNABORTED: 'ABORTED',  # this host's network is broken
    # broken pipes: we never write on the socket so should never happen
}


TIMEOUT = 15.0


async def handle_connection(host, port, proto=socket.IPPROTO_TCP, name=''):
    beginning_t = time()
    status = 'OK'
    try:
        c = await asyncio.wait_for(
            asyncio.open_connection(
                host,
                port,
                family=socket.AF_INET,
                proto=proto,
            ),
            timeout=TIMEOUT,
        )
        c[1].close()
    except asyncio.TimeoutError:
        # also OsError can be timeout
        status = 'TIMEDOUT'
    except OSError as ex:
        status = STATUSES[ex.errno]
        if ex.errno == errno.ETIMEDOUT:
            sys.stdout.write('actual oserror timeout')
            sys.stdout.flush()
    return {
        'status': status,
        'elapsed': time() - beginning_t,
        'host': host,
        'port': port,
        'proto': proto,
        'name': name,
    }


def ip_generator():
    ip = socket.inet_ntoa(struct.pack('>I', random.randint(1, 0xffffffff)))
    splitted_ip = [int(n) for n in ip.split('.')]
    if (
            splitted_ip[0] == 0
            or 255 in splitted_ip
            or splitted_ip[0] >= 224
            or splitted_ip[0] == 127
    ):
        yield next(ip_generator())
    yield ip


class SocketGroup(asyncio.TaskGroup):
    def __init__(self):
        super().__init__()
        for v in STATUSES.values():
            setattr(self, v, 0)
        self.elapsed = {}

    def _on_task_done(self, task):
        res = task.result()
        # python has a nice ++ operator
        setattr(self, res['status'], (getattr(self, res['status']) + 1))
        if res['elapsed'] < .1:
            self.elapsed.setdefault(.1, []).append(res['elapsed'])
        elif res['elapsed'] < .3:
            self.elapsed.setdefault(.3, []).append(res['elapsed'])
        elif res['elapsed'] < .5:
            self.elapsed.setdefault(.5, []).append(res['elapsed'])
        elif res['elapsed'] < 1:
            self.elapsed.setdefault(1, []).append(res['elapsed'])
        elif res['elapsed'] < 1.2:
            self.elapsed.setdefault(1.2, []).append(res['elapsed'])
        elif res['elapsed'] < 2.9:
            self.elapsed.setdefault(2.9, []).append(res['elapsed'])
        elif res['elapsed'] < 3.4:
            self.elapsed.setdefault(3.4, []).append(res['elapsed'])
        elif res['elapsed'] < 5:
            self.elapsed.setdefault(5, []).append(res['elapsed'])
        else:
            self.elapsed.setdefault('+', []).append(res['elapsed'])
        return super()._on_task_done(task)


async def main():
    print(f'with timeout set to {TIMEOUT} seconds')
    while True:
        async with SocketGroup() as tg:
            for i in range(512):
                # cannot exceed 1024 otherwise too
                # many file descriptors are open!
                ip = next(ip_generator())
                tg.create_task(
                    handle_connection(
                        ip,
                        80,
                        name=str(i),
                    ),
                )
        times = [f"{s}: {len(v)}" for s, v in tg.elapsed.items()]
        print(f'OK={tg.OK} TIMEDOUT={tg.TIMEDOUT} elapsed={times}')

asyncio.run(main())

```
execution example
```
$ python connection_test.py
with timeout set to 15.0 seconds
OK=511 TIMEDOUT=1 elapsed=['0.1: 161', '1.2: 133', '3.4: 132', '+: 86']
OK=512 TIMEDOUT=0 elapsed=['0.1: 130', '0.3: 30', '1.2: 197', '3.4: 132', '+: 23']
OK=511 TIMEDOUT=1 elapsed=['0.1: 92', '0.3: 103', '1.2: 243', '3.4: 73', '+: 1']
OK=512 TIMEDOUT=0 elapsed=['0.1: 148', '0.3: 23', '1.2: 141', '3.4: 130', '+: 70']
OK=512 TIMEDOUT=0 elapsed=['0.1: 157', '0.3: 14', '1.2: 131', '3.4: 135', '+: 75']
OK=511 TIMEDOUT=1 elapsed=['0.1: 153', '0.3: 46', '1.2: 214', '3.4: 98', '+: 1']
OK=512 TIMEDOUT=0 elapsed=['0.1: 136', '0.3: 39', '1.2: 202', '3.4: 124', '+: 11']
OK=512 TIMEDOUT=0 elapsed=['0.1: 107', '0.3: 74', '1.2: 134', '3.4: 127', '+: 70']
OK=512 TIMEDOUT=0 elapsed=['0.1: 122', '0.3: 56', '1.2: 131', '3.4: 130', '+: 73']
OK=511 TIMEDOUT=1 elapsed=['0.1: 77', '0.3: 84', '1.2: 139', '3.4: 137', '+: 75']
OK=512 TIMEDOUT=0 elapsed=['0.1: 146', '0.3: 25', '1.2: 179', '3.4: 130', '+: 32']
OK=512 TIMEDOUT=0 elapsed=['0.1: 138', '0.3: 36', '1.2: 129', '3.4: 129', '+: 80']
OK=512 TIMEDOUT=0 elapsed=['0.1: 123', '0.3: 49', '1.2: 277', '3.4: 63']
OK=512 TIMEDOUT=0 elapsed=['0.1: 157', '0.3: 2', '1.2: 138', '3.4: 129', '+: 86']
OK=512 TIMEDOUT=0 elapsed=['0.1: 142', '0.3: 25', '1.2: 133', '3.4: 129', '+: 83']
OK=512 TIMEDOUT=0 elapsed=['0.1: 112', '0.3: 49', '1.2: 255', '3.4: 96']
OK=512 TIMEDOUT=0 elapsed=['0.1: 161', '0.3: 3', '1.2: 148', '3.4: 133', '+: 67']
OK=512 TIMEDOUT=0 elapsed=['0.1: 145', '0.3: 7', '1.2: 143', '3.4: 125', '+: 92']
OK=512 TIMEDOUT=0 elapsed=['0.1: 117', '0.3: 39', '1.2: 131', '3.4: 129', '+: 96']
OK=512 TIMEDOUT=0 elapsed=['0.1: 122', '0.3: 74', '1.2: 162', '3.4: 141', '+: 13']
OK=512 TIMEDOUT=0 elapsed=['0.1: 153', '1.2: 130', '3.4: 128', '+: 101']
OK=512 TIMEDOUT=0 elapsed=['0.1: 168', '1.2: 247', '3.4: 97']
OK=512 TIMEDOUT=0 elapsed=['0.1: 105', '0.3: 105', '1.2: 196', '3.4: 106']
OK=512 TIMEDOUT=0 elapsed=['0.1: 157', '1.2: 137', '3.4: 127', '+: 91']
OK=512 TIMEDOUT=0 elapsed=['0.1: 162', '1.2: 222', '3.4: 117', '+: 11']
OK=512 TIMEDOUT=0 elapsed=['0.1: 158', '1.2: 151', '3.4: 121', '+: 82']
OK=512 TIMEDOUT=0 elapsed=['0.1: 162', '0.3: 1', '1.2: 130', '3.4: 128', '+: 91']
OK=512 TIMEDOUT=0 elapsed=['0.1: 148', '1.2: 257', '3.4: 107']
OK=512 TIMEDOUT=0 elapsed=['0.1: 162', '1.2: 165', '3.4: 131', '+: 54']
OK=512 TIMEDOUT=0 elapsed=['0.1: 87', '0.3: 69', '1.2: 134', '3.4: 120', '+: 102']
OK=512 TIMEDOUT=0 elapsed=['0.1: 130', '0.3: 26', '1.2: 253', '3.4: 103']
OK=512 TIMEDOUT=0 elapsed=['0.1: 135', '0.3: 21', '1.2: 143', '3.4: 128', '+: 85']
OK=512 TIMEDOUT=0 elapsed=['0.1: 92', '0.3: 69', '1.2: 139', '3.4: 130', '+: 82']
OK=512 TIMEDOUT=0 elapsed=['0.1: 125', '0.3: 38', '1.2: 144', '3.4: 130', '+: 75']
OK=512 TIMEDOUT=0 elapsed=['0.1: 160', '1.2: 142', '3.4: 125', '+: 85']
OK=512 TIMEDOUT=0 elapsed=['0.1: 129', '0.3: 47', '1.2: 136', '3.4: 132', '+: 68']
OK=512 TIMEDOUT=0 elapsed=['0.1: 107', '0.3: 63', '1.2: 135', '3.4: 130', '+: 77']
OK=512 TIMEDOUT=0 elapsed=['0.1: 135', '0.3: 32', '1.2: 128', '3.4: 130', '+: 87']
OK=512 TIMEDOUT=0 elapsed=['0.1: 161', '1.2: 142', '3.4: 123', '+: 86']
OK=512 TIMEDOUT=0 elapsed=['0.1: 132', '0.3: 22', '1.2: 181', '3.4: 126', '+: 51']
OK=512 TIMEDOUT=0 elapsed=['0.1: 85', '0.3: 70', '1.2: 234', '3.4: 116', '+: 7']
OK=503 TIMEDOUT=9 elapsed=['0.1: 135', '1.2: 127', '3.4: 130', '+: 120']
OK=512 TIMEDOUT=0 elapsed=['0.1: 153', '0.3: 12', '1.2: 208', '3.4: 126', '+: 13']
OK=511 TIMEDOUT=1 elapsed=['0.1: 121', '0.3: 49', '1.2: 240', '3.4: 101', '+: 1']
OK=512 TIMEDOUT=0 elapsed=['0.1: 101', '0.3: 78', '1.2: 133', '3.4: 126', '+: 74']
OK=512 TIMEDOUT=0 elapsed=['0.1: 153', '0.3: 7', '1.2: 133', '3.4: 131', '+: 88']
OK=512 TIMEDOUT=0 elapsed=['0.1: 64', '0.3: 197', '1.2: 164', '3.4: 87']
OK=510 TIMEDOUT=2 elapsed=['0.1: 157', '1.2: 247', '3.4: 106', '+: 2']
OK=508 TIMEDOUT=4 elapsed=['0.1: 106', '0.3: 52', '1.2: 127', '3.4: 128', '+: 99']
OK=512 TIMEDOUT=0 elapsed=['0.1: 140', '0.3: 23', '1.2: 227', '3.4: 117', '+: 5']
OK=512 TIMEDOUT=0 elapsed=['0.1: 156', '0.3: 10', '1.2: 130', '3.4: 130', '+: 86']
OK=511 TIMEDOUT=1 elapsed=['0.1: 146', '0.3: 16', '1.2: 140', '3.4: 123', '+: 87']
OK=512 TIMEDOUT=0 elapsed=['0.1: 124', '0.3: 46', '1.2: 137', '3.4: 132', '+: 73']
OK=512 TIMEDOUT=0 elapsed=['0.1: 109', '0.3: 50', '1.2: 138', '3.4: 134', '+: 81']
OK=512 TIMEDOUT=0 elapsed=['0.1: 77', '0.3: 165', '1.2: 194', '3.4: 76']
OK=512 TIMEDOUT=0 elapsed=['0.1: 146', '0.3: 9', '1.2: 246', '3.4: 110', '+: 1']
OK=511 TIMEDOUT=1 elapsed=['0.1: 144', '0.3: 14', '1.2: 140', '3.4: 129', '+: 85']
OK=512 TIMEDOUT=0 elapsed=['0.1: 134', '0.3: 79', '1.2: 135', '3.4: 134', '+: 30']
OK=512 TIMEDOUT=0 elapsed=['0.1: 162', '1.2: 259', '3.4: 91']
OK=510 TIMEDOUT=2 elapsed=['0.1: 159', '1.2: 181', '3.4: 130', '+: 42']
OK=512 TIMEDOUT=0 elapsed=['0.1: 162', '1.2: 228', '3.4: 118', '+: 4']
OK=510 TIMEDOUT=2 elapsed=['0.1: 153', '1.2: 130', '3.4: 129', '+: 100']
OK=512 TIMEDOUT=0 elapsed=['0.1: 164', '1.2: 133', '3.4: 124', '+: 91']
OK=512 TIMEDOUT=0 elapsed=['0.1: 95', '0.3: 107', '1.2: 156', '3.4: 134', '+: 20']
OK=512 TIMEDOUT=0 elapsed=['0.1: 146', '0.3: 7', '1.2: 172', '3.4: 132', '+: 55']
OK=512 TIMEDOUT=0 elapsed=['0.1: 122', '0.3: 40', '1.2: 207', '3.4: 137', '+: 6']
OK=511 TIMEDOUT=1 elapsed=['0.1: 127', '0.3: 40', '1.2: 196', '3.4: 122', '+: 27']
OK=511 TIMEDOUT=1 elapsed=['0.1: 132', '0.3: 87', '1.2: 257', '3.4: 35', '+: 1']
OK=512 TIMEDOUT=0 elapsed=['0.1: 158', '0.3: 10', '1.2: 146', '3.4: 130', '+: 68']
OK=512 TIMEDOUT=0 elapsed=['0.1: 166', '0.3: 7', '1.2: 132', '3.4: 131', '+: 76']
OK=512 TIMEDOUT=0 elapsed=['0.1: 160', '1.2: 258', '3.4: 94']
OK=512 TIMEDOUT=0 elapsed=['0.1: 133', '0.3: 46', '1.2: 208', '3.4: 117', '+: 8']
OK=512 TIMEDOUT=0 elapsed=['0.1: 166', '0.3: 2', '1.2: 134', '3.4: 133', '+: 77']
OK=512 TIMEDOUT=0 elapsed=['0.1: 137', '0.3: 61', '1.2: 222', '3.4: 92']
OK=512 TIMEDOUT=0 elapsed=['0.1: 96', '0.3: 90', '1.2: 268', '3.4: 58']
OK=512 TIMEDOUT=0 elapsed=['0.1: 109', '0.3: 57', '1.2: 266', '3.4: 80']
OK=512 TIMEDOUT=0 elapsed=['0.1: 147', '0.3: 14', '1.2: 133', '3.4: 129', '+: 89']
OK=512 TIMEDOUT=0 elapsed=['0.1: 123', '0.3: 46', '1.2: 263', '3.4: 80']
OK=512 TIMEDOUT=0 elapsed=['0.1: 159', '1.2: 136', '3.4: 114', '+: 103']
OK=512 TIMEDOUT=0 elapsed=['0.1: 130', '0.3: 31', '1.2: 259', '3.4: 92']
OK=512 TIMEDOUT=0 elapsed=['0.1: 132', '0.3: 26', '1.2: 143', '3.4: 123', '+: 88']
OK=511 TIMEDOUT=1 elapsed=['0.1: 120', '0.3: 46', '1.2: 138', '3.4: 137', '+: 71']
OK=512 TIMEDOUT=0 elapsed=['0.1: 159', '1.2: 221', '3.4: 123', '+: 9']
OK=511 TIMEDOUT=1 elapsed=['0.1: 101', '0.3: 63', '1.2: 124', '3.4: 132', '+: 92']
OK=512 TIMEDOUT=0 elapsed=['0.1: 157', '0.3: 5', '1.2: 268', '3.4: 82']
OK=511 TIMEDOUT=1 elapsed=['0.1: 103', '0.3: 63', '1.2: 264', '3.4: 81', '+: 1']
OK=512 TIMEDOUT=0 elapsed=['0.1: 64', '0.3: 99', '1.2: 272', '3.4: 77']
OK=512 TIMEDOUT=0 elapsed=['0.1: 111', '0.3: 47', '1.2: 125', '3.4: 134', '+: 95']
OK=512 TIMEDOUT=0 elapsed=['0.1: 98', '0.3: 83', '1.2: 169', '3.4: 129', '+: 33']
OK=511 TIMEDOUT=1 elapsed=['0.1: 153', '1.2: 216', '3.4: 125', '+: 18']
OK=512 TIMEDOUT=0 elapsed=['0.1: 102', '0.3: 56', '1.2: 246', '3.4: 108']
OK=511 TIMEDOUT=1 elapsed=['0.1: 112', '0.3: 67', '1.2: 130', '3.4: 130', '+: 73']
OK=512 TIMEDOUT=0 elapsed=['0.1: 75', '0.3: 94', '1.2: 131', '3.4: 138', '+: 74']
OK=512 TIMEDOUT=0 elapsed=['0.1: 113', '0.3: 48', '1.2: 162', '3.4: 130', '+: 59']
OK=512 TIMEDOUT=0 elapsed=['0.1: 82', '0.3: 81', '1.2: 131', '3.4: 121', '+: 97']
OK=512 TIMEDOUT=0 elapsed=['0.1: 116', '0.3: 46', '1.2: 135', '3.4: 131', '+: 84']
OK=512 TIMEDOUT=0 elapsed=['0.1: 111', '0.3: 49', '1.2: 263', '3.4: 89']
OK=512 TIMEDOUT=0 elapsed=['0.1: 153', '1.2: 216', '3.4: 119', '+: 24']
OK=512 TIMEDOUT=0 elapsed=['0.1: 80', '0.3: 127', '1.2: 137', '3.4: 126', '+: 42']
OK=512 TIMEDOUT=0 elapsed=['0.1: 95', '0.3: 71', '1.2: 125', '3.4: 134', '+: 87']
```

As you can see there are no responses in the **0.3-1.0** range, neither in the **1.2-2.9** or **3.4-5** ranges

Can you explain what's going on?

Thanks

# Your environment

- CPython versions tested on: 3.11.0 (and 3.10 using gather instead of TaskGroup)
- Operating system and architecture: archlinux x86_64
- cpu: i7-1165G7
- network controllers: RTL8111/8168/8411 and intel AX201
- two network connections from different ISP

**Comment by gvanrossum** [comment]
I'm sorry, I'm not going to try to debug that much code. I'm not even sure what you're saying the problem is. Just that the distribution of timeouts is not what you expected? Why should that point to a bug in asyncio?

**Comment by CastixGitHub** [comment]
The distribution of timeouts is right (always about 15 seconds)
The distribution of OK results is my concern, it seems the network becomes unusable in those ranges (just for the script. eg: playing something with sox works without holes. `while true; do time nc -zv 9.9.9.9 53; done` keeps working)

with some small changes i've made some plotting with 10ms groups

![image](https://user-images.githubusercontent.com/21062669/209203907-ab46e820-8f0b-4465-8b67-cbc4cd4928f2.png)

here's the diff for reproducibility:
```diff
$ diff connection_test_0.py connection_test.py
82,99c82,83
<         if res['elapsed'] < .1:
<             self.elapsed.setdefault(.1, []).append(res['elapsed'])
<         elif res['elapsed'] < .3:
<             self.elapsed.setdefault(.3, []).append(res['elapsed'])
<         elif res['elapsed'] < .5:
<             self.elapsed.setdefault(.5, []).append(res['elapsed'])
<         elif res['elapsed'] < 1:
<             self.elapsed.setdefault(1, []).append(res['elapsed'])
<         elif res['elapsed'] < 1.2:
<             self.elapsed.setdefault(1.2, []).append(res['elapsed'])
<         elif res['elapsed'] < 2.9:
<             self.elapsed.setdefault(2.9, []).append(res['elapsed'])
<         elif res['elapsed'] < 3.4:
<             self.elapsed.setdefault(3.4, []).append(res['elapsed'])
<         elif res['elapsed'] < 5:
<             self.elapsed.setdefault(5, []).append(res['elapsed'])
<         else:
<             self.elapsed.setdefault('+', []).append(res['elapsed'])
---
>         ten_ms_grp = self.elapsed.setdefault(float('%.3f' % res['elapsed']), [])
>         ten_ms_grp.append(res['elapsed'])
119a104,122
>         print(tg.elapsed)
>         if True:
>             from matplotlib import pyplot, colors
>             fig = pyplot.figure(figsize=(10, 6), dpi=300)
>             pyplot.title('10ms groups')
>             pyplot.xlabel('time')
>             pyplot.ylabel('count')
>             _colors = [
>                 random.choice(list(colors.cnames.values()))
>                 for _ in range(len(tg.elapsed.keys()))
>             ]
>             pyplot.bar(
>                 tg.elapsed.keys(),
>                 [len(v) for v in tg.elapsed.values()],
>                 width=0.001,
>                 color=_colors,
>             )
>             fig.savefig('fig.png')
>             
```

it seems asyncio just waits too much, the resource it is waiting for is free, the cpu usage is about 0%
is there some backoff mechanism on busy resources (trying to guess by looking at the disposition)?

**Comment by gvanrossum** [comment]
Could you try to explain in plain English what the code is doing and what you are measuring? I am a bear with very little brain and I just can’t follow your code.

**Comment by gvanrossum** [comment]
Maybe it’s a property of random IP addresses? Maybe you could read the source code?

**Comment by CastixGitHub** [comment]
sorry @gvanrossum about the lack of context

the code is just creating batches of cooperative tasks that check the availability of a service.
ip addresses were picked randomly, all the checks are on tcp protocol and port 80

each coroutine measures the time elapsed between the start of the connection and it's end
the connection is immediatly closed without any data transmission on the channel

the distributions is not related to the ip geolocation, if you set a fixed ip, say 9.9.9.9, the distribution doesn't change.

setting the ip to 127.0.0.1 or other devices in the same lan the issue is not shown, the bar plot is as expected (i also increased batch size)
![image](https://user-images.githubusercontent.com/21062669/209215380-1a0fe68c-4c64-498a-b017-c3cb6af55419.png)

doing a traceroute and picking the next hop after my home gateway, the issue is present
 
the total absence of answers in those time ranges made it clear it wasn't a related to the target geolocation

here's a shrinked version of the code, this should be easier to follow
```python
# © 2022 Vincenzo Castiglia <vincenzo.castiglia@par-tec.it> 3F2AD937F3D77362
import asyncio
import socket
from time import time


TIMEOUT = 15.0
BATCH_SIZE = 512
IP = '138.197.63.241'
PORT = 80


async def handle_connection(ip, port):
    beginning_t = time()
    try:
        c = await asyncio.wait_for(
            asyncio.open_connection(ip, port),
            timeout=TIMEOUT,
        )
        c[1].close()
    except (asyncio.TimeoutError, OSError):
        pass
    return time() - beginning_t


class SocketGroup(asyncio.TaskGroup):
    def __init__(self):
        super().__init__()
        self.elapsed = {}

    def _on_task_done(self, task):
        # store the elapsed time grouping them by 10 milliseconds
        elapsed = task.result()
        self.elapsed.setdefault(float('%.3f' % elapsed), []).append(elapsed)
        return super()._on_task_done(task)


async def main():
    while True:
        async with SocketGroup() as tg:
            for i in range(BATCH_SIZE):
                tg.create_task(handle_connection(IP, PORT))
        # print the count for each group
        print({grp: len(_list) for grp, _list in tg.elapsed.items()})


asyncio.run(main())

```
will output lines similar to
```
{0.03: 1, 0.034: 2, 0.038: 5, 0.043: 6, 0.059: 9, 0.06: 3, 0.061: 14, 0.062: 1, 0.065: 8, 0.066: 12, 0.068: 3, 0.076: 1, 0.08: 4, 0.078: 1, 0.081: 5, 0.084: 16, 0.086: 14, 0.09: 12, 0.089: 2, 0.091: 4, 0.093: 25, 0.099: 8, 0.1: 25, 0.103: 5, 0.104: 21, 0.105: 4, 1.037: 1, 1.053: 4, 1.054: 9, 1.055: 1, 1.056: 6, 1.057: 5, 1.06: 10, 1.061: 13, 1.063: 7, 1.067: 11, 1.068: 1, 1.069: 8, 1.07: 9, 1.076: 4, 1.077: 8, 1.078: 1, 1.079: 8, 1.08: 1, 1.082: 4, 1.083: 15, 1.085: 12, 1.086: 9, 1.084: 1, 1.087: 1, 1.088: 3, 1.089: 4, 1.09: 12, 1.091: 10, 1.094: 5, 1.095: 8, 1.096: 16, 1.102: 14, 1.103: 9, 1.107: 7, 3.24: 1, 3.258: 6, 3.262: 13, 3.264: 1, 3.265: 22, 3.27: 19, 3.271: 2}
```
Thanks for the help

**Comment by CastixGitHub** [comment]
> Why should that point to a bug in asyncio?

I don't know, it can also be a bug in kernel's epoll. but still the results i get from python are unexpected

is a comparative rewrite in another language helpful or just confusing?

**Comment by gvanrossum** [comment]
Maybe first rerun with uvloop?

That you don’t see the effect with local host makes me think it is not in Python?

**Comment by gvanrossum** [comment]
I think there's stuff lingering in the OS kernel that slows you down. On my Mac, your last version gives a pretty nice distribution the first time, then everything is slowed down by about a second or more. But if I just wait several seconds between runs I get the expected (initial) distribution every time.

Please close this issue if you agree.

**Comment by CastixGitHub** [comment]
<s>

- uvloop shows the same results

- in a scenario where you open 470 tcp sockets every second, after a minute all the available ports will be used (TIME_WAIT). This is not causing the issue

- a faster connection seems to help (test with a 2G cellular network)
</s>

This is just RTO, the backoff is explained in [rfc6298](https://www.rfc-editor.org/rfc/rfc6298)

**Comment by gvanrossum** [comment]
Thanks for confirming!
